Refactor extract text job to use more pyspark.sql

jodizzle opened the pull request:
- Rework `corpus/jobs/extract_text.py` to be based on UDFs and dataframes rather than RDDs.
- Remove the `DocumentText.from_document` method, transplanting it into a helper method within `extract_text.py`.

This job should probably be run and validated before merging.  Some more thorough testing wouldn't hurt either.

One issue brought up by this new approach is error handling.  In the RDD approach, any errors are caught and logged via the `try_or_log` decorator, and then filtered out before saving the tables.  As far as I know, this won't work with the dataframes approach.  Consequently, if the `Document.data` field of any row is null, or if `osp_text_extractor.bytes_to_text` throws an error, the job will crash (or I'm fairly certain it will, anyway).  Should the error handling for this type of approach be done in a case-by-case (bespoke) way, or is there some more generic approach similar to `try_or_log`?

jodizzle commented:
I just realized that trying to place logging statements inside a function being UDF'd can be problematic -- Spark will apparently give serialization errors.  The only reason there aren't such errors currently is because I actually forgot to import `osp_pipeline.log`.  Another thing to fixup.

wearpants commented on osp_pipeline/corpus/jobs/extract_text.py:
```diff
@@ -1,19 +1,67 @@
+        ])
+    )
+
+    getMD5 = udf(
```
You should be able to do this with spark built-in md5 function instead of Python call

jodizzle replied:
Yeah, I realized too late that Spark has built-in md5 stuff.  Given that, I guess we can probably (almost) remove the `osp_pipeline.corpus.utils.get_md5` function.

wearpants commented on osp_pipeline/corpus/jobs/extract_text.py:
```diff
@@ -1,19 +1,67 @@
+        docs
+        .withColumn("output", extractText('data', 'id'))
+        .select("id", "output.mime_type", "output.text")
+        .withColumn('text_md5', getMD5('text'))
```
Youd replace this UDF with a native spark sql expression

wearpants commented on osp_pipeline/corpus/jobs/extract_text.py:
```diff
@@ -1,19 +1,67 @@
+
+    Returns: (mime_type <str>, text <str>)
+    """
+    # TODO: Better exception handling?
```
I'd wrap the entire function thing in try_or_none, and then wrap that in a little if-else function that guarantees to return a 2-tuple of (mimetype, data) or (none, none)

jodizzle replied:
Do you mean `try_or_log`?  Unfortunately, I don't think I can use it at the moment, because of the issue with serializing python loggers (see my comment above, the usual error involves something like `can't pickle _thread.RLock objects`).  I'm looking into some solutions.

wearpants commented on osp_pipeline/corpus/jobs/extract_text.py:
```diff
@@ -1,19 +1,67 @@
+    result = (
+        docs
+        .withColumn("output", extractText('data', 'id'))
+        .select("id", "output.mime_type", "output.text")
```
I'd filter out with output.isNotNull() at this point

wearpants commented:
Yeah. We can modify try_or_log to always get a new logger in each call with
logging.getLogger() instead of passing it implicitly in a closure  that
should solve your problem

On Jan 2, 2018 10:57 AM, "jodizzle" <notifications@github.com> wrote:

> *@jodizzle* commented on this pull request.
> ------------------------------
>
> In osp_pipeline/corpus/jobs/extract_text.py
> <https://github.com/opensyllabus/osp-pipeline/pull/107#discussion_r159253990>
> :
>
> > -def extract_text(row):
> -    return DocumentText.from_document(row)
> +def extract_text(doc_data, doc_id):
> +    """Extract text from Document bytes.
> +
> +    Text is extracted via `osp_text_extractor.bytes_to_text`.  If there
> +    are errors during text extraction, a warning is logged that includes
> +    a reference to 'doc_id'.
> +
> +    Args:
> +        doc_data (bytes): Bytes to extract text from.
> +        doc_id (str): id of the Document.
> +
> +    Returns: (mime_type <str>, text <str>)
> +    """
> +    # TODO: Better exception handling?
>
> Do you mean try_or_log? Unfortunately, I don't think I can use it at the
> moment, because of the issue with serializing python loggers (see my
> comment above, the usual error involves something like can't pickle
> _thread.RLock objects). I'm looking into some solutions.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/opensyllabus/osp-pipeline/pull/107#discussion_r159253990>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAe2APo7P7w0Z-FNBApnDBNKSs2xMSb9ks5tGlHSgaJpZM4RQT0z>
> .
>

jodizzle commented:
@wearpants I think I addressed all your comments -- `try_or_log`, Pyspark's built-in `md5`, and filtering null output.  If it looks good, I propose we try running it, and then merge once we've confirmed everything is working.

wearpants commented:
Sounds good.
